### PR TITLE
PB-4302: Implement automatic resource transformation for Kubevirt

### PR DESCRIPTION
**Pull request by vsundarraj-px**
**What type of PR is this?**
> Uncomment only one and also add the corresponding label in the PR:
>bug
feature
>improvement
>cleanup
>api-change
>design
>documentation
>failing-test
>unit-test
>integration-test

**What this PR does / why we need it**:
This changes fixes the issue while restoring kubevirt VMs with networking configurations and also with DatavolumeTemplates.

More details in https://portworx.atlassian.net/browse/PB-4302

**Does this PR change a user-facing CRD or CLI?**:

NO
**Is a release note needed?**:

```release-note
Issue: https://portworx.atlassian.net/browse/PB-4302
User Impact: None
Resolution

```

**Does this change need to be cherry-picked to a release branch?**:

NO

### CODE REVIEW

File: pkg/resourcecollector/virtualmachine.go [hunk 1]
```diff
@@ -5,6 +5,102 @@ import (
...
+ * VM resources.
+ */
+func (r *ResourceCollector) prepareVirtualMachineForApply(
+	object runtime.Unstructured,
+	namespaces []string,
+) error {
+
+	// Remove macAddress as it creates macAddress conflict while restore
+	// to the same cluster to different namespace while the source VM is up and
+	path := []string{"spec", "template", "spec", "domain", "devices", "interfaces"}
```
Comment (prashanthpx) [inline pkg/resourcecollector/virtualmachine.go:91]: Can we have the path as `.` separated for easier understanding. eg `spec.template.spec.domain.devices.interfaces`

File: pkg/resourcecollector/virtualmachine.go [hunk 2]
```diff
@@ -5,6 +5,102 @@ import (
...
+        if er != nil {
+		return er
+        }
+        found := false
+        interf := make([]interface{}, 0, 1)
+        for _, val := range data {
+		switch typeval := val.(type) {
+		case map[string]interface{}:
+			newMap := make(map[string]interface{})
+			parseMap(typeval, newMap, key, &found)
```
Comment (prashanthpx) [inline pkg/resourcecollector/virtualmachine.go:64]: Here would `unstructured.SetNestedSlice` unset entire interfaces?  I was thinking we want to unset entire `interface` field. 
Can you add exmaple comment for this case?

**Comment by prashanthpx** [comment]
@diptiranjanpx @siva-portworx Can you please review this PR?

File: pkg/resourcecollector/virtualmachine.go [hunk 3]
```diff
@@ -5,6 +5,102 @@ import (
...
+	}
+}
+
+/*
+ * Path parsing function for virtualMachine template.
+ */
+func transformPath(object runtime.Unstructured, path []string, key string) error {
+
+        content := object.UnstructuredContent()
+        data, _, er := unstructured.NestedSlice(content, path...)
```
Comment (diptiranjanpx) [inline pkg/resourcecollector/virtualmachine.go:54]: nit: err.

File: pkg/resourcecollector/virtualmachine.go [hunk 4]
```diff
@@ -5,6 +5,102 @@ import (
...
+		}
+		output[key] = val
+	}
+	return output
+}
+
+/*
+ * Utility function for virtualMachine template parsing logic
+ */
+func parseMap(input map[string]interface{}, output map[string]interface{}, key string, is_key_present *bool) {
```
Comment (diptiranjanpx) [inline pkg/resourcecollector/virtualmachine.go:29]: what is the use of argument `key` here?

File: pkg/resourcecollector/virtualmachine.go (hunk 2 above)
Comment (vsundarraj-px) [reply pkg/resourcecollector/virtualmachine.go]: we will unset entire slice.

File: pkg/resourcecollector/virtualmachine.go [hunk 5]
```diff
@@ -1,10 +1,101 @@
...
+        data, _, err := unstructured.NestedSlice(content, path...)
+        if err != nil {
+		return err
+        }
+        found := false
+        interf := make([]interface{}, 0, 1)
+        for _, val := range data {
+		switch typeval := val.(type) {
+		case map[string]interface{}:
+			newMap := make(map[string]interface{})
```
Comment (prashanthpx) [inline pkg/resourcecollector/virtualmachine.go:59]: As discussed, can we try `RemoveNestedField()` and passing macAddress field

Comment (vsundarraj-px) [reply pkg/resourcecollector/virtualmachine.go:59]: Tried and this doesnt work. Since macAddress is not a nestedField.

**Review by prashanthpx — APPROVED** [review]
LGTM. Lets hold on for merging as we need to decide for which stork release this will be picked up

**Comment by vsundarraj-px** [comment]
Screenshot 2023-08-29 at 9 44 55 AM

Yaml after restoring it to new namespace px-restored.. with macAddress changed and DatavolumeTemplate and DataVolume changed to associated PVC

**Comment by vsundarraj-px** [comment]
Screenshot 2023-09-05 at 6 36 49 AM

Yaml before taking backup

**Comment by vsundarraj-px** [comment]
Screenshot 2023-08-29 at 9 44 11 AM
Yaml Before px-backup

**Comment by vsundarraj-px** [comment]
rebuild

**Comment by adityadani** [comment]
Why is a RL note not needed for this? Any change that impacts functionality should have a RL note.

**Comment by adityadani** [comment]
Looks like the changes from this PR are already merged as part of #1502. @vsundarraj-px Do you still need this PR open?

**Comment by vsundarraj-px** [comment]
This is duplicate of PF #1502 created due to fork PR. Hence closing.

**Pull request closed without merging**